src/Swarm/Game/Step.hs

```diff
@@ -774,8 +774,28 @@ execConst c vs s k = do
         flagRedraw
         return $ Out VUnit s k
       _ -> badConst
-    Grab -> doGrab False
-    Harvest -> doGrab True
+    Grab -> doGrab Grab'
+    Harvest -> doGrab Harvest'
+    Swap -> case vs of
+      [VText name] -> do
+        inv <- use robotInventory
```

Comment:
Some of this code seems like it's duplicated from elsewhere; are there any additional opportunities for factoring out common utility functions?

Reply:
Yes, it is what `place` does, with some changes and some omissions.

I thought about refactoring this further, but it is not as similar to `place` as `grab` is to `harvest`, so I do not think it would make the code clearer.

Reply:
OK, I found a small thing to improve. :grin: